### ISSUE TITLE
Change Request: Config to Disallow Disable

### ISSUE THREAD

**Issue by Brian-Bartels** [eslint/eslint#15631]
### ESLint version

8.0.0

### What problem do you want to solve?

ESLint currently allows users to disable rules. In large codebases with MANY Prs, some users disable lint rules without really knowing what they are doing. This slips by PRs because reviewers sometimes don't actually review ¯\_(ツ)_/¯

There exists rules such as `no-abusive-eslint-disable` but that in and of itself can be disabled

### What do you think is the correct solution?

I'd like ESLint to allow users to configure certain rules to be exempt from all comment style disables

eslintrc: `"noExtraSemi": ["error", "dissalowDisable"]`

so in code
```
// eslint-disable-next-line noExtraSemi
const foo = 5;;
```
Would still result in an error

### Participation

- [ ] I am willing to submit a pull request for this change.

**Comment by nzakas** [comment]
Would noInlineConfig solve this problem for you?
https://eslint.org/docs/user-guide/configuring/rules#disabling-inline-comments

**Comment by Brian-Bartels** [comment]
@nzakas That would work for some. But for us where we have custom lint rules to enforce security or other critical pieces. Disabling some rules is okay where disabling other is not okay

**Comment by Pzixel** [comment]
Just for a reference, in rust there are 3 levels of enabling lints/rules/...
allow(x) - allow violation of rule x
deny(x) - forbids violation of rule x
forbid(x) - forbids violation of rule x and makes it hard error to do allow(x)

See [link](https://doc.rust-lang.org/rustc/lints/levels.html) for details.

**Comment by Brian-Bartels** [comment]
ohhhh @Pzixel I do like that implementation much better than what I had proposed

**Comment by nzakas** [comment]
I’m on the fence as to whether there’s enough of a use case here to rationalize the additional complexity to the config system. At the moment, the tradeoff here doesn’t seem worth it.  I’m curious what other team members think. 

As a note: we aren’t making any further changes to the current config system as we shift to a new config system, so any potential changes would be a bit in the future.

**Comment by Brian-Bartels** [comment]
As evidence for people being bad PR reviewers following lines are in a repo that I work on over 30 times. And yes it's ridiculous that this get's checked in. That's just the nature of incredibly large repos though :(

```
// eslint-disable-next-line foo/no-abusive-eslint-disable
// eslint-disable-next-line foo/no-eslint-disable
// eslint-disable-next-line foo/no-foo-disable-with-exceptions
```

**Comment by mdjermanovic** [comment]
I wouldn't be opposed to adding this feature to the core if there are no alternatives. In my understanding, some plugin rules like [`eslint-comments/no-restricted-disable`](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-restricted-disable.html) provide the same functionality, but the only problem is that those rules can be disabled as well. Would it work for you to run eslint again with the [`--no-inline-config`](https://eslint.org/docs/user-guide/command-line-interface#--no-inline-config) flag and a config file that has only rules related to eslint comments enabled?

**Comment by ljharb** [comment]
Disabling all inline config doesn’t seem like it solves the problem - some rules *should* be disable-able with inline config; some should not.

**Comment by mdjermanovic** [comment]
> Disabling all inline config doesn’t seem like it solves the problem - some rules _should_ be disable-able with inline config; some should not.

I agree with that, the solution I suggested is another run only with rules that provide this functionality, assuming that the only problem is how to prevent disabling them.

```
npx eslint . --no-inline-config --no-eslintrc -c .eslintrc2.js
```

```js
//--------- .eslintrc2.js -----------

module.exports = {
    plugins: ["eslint-comments"],
    rules: {
        "eslint-comments/no-restricted-disable": ["error", "no-undef" /*, ... */]
    }
};
```

**Comment by Brian-Bartels** [comment]
@mdjermanovic One concern with that workaround. How would it integrate into VSCode? I worry that people would disable the rule expecting it to pass PR build step. But then we'd sneak in a second run of ESlint and fail their builds unexpectedly.

I do also think that that workaround falls into the "slightly hacky" category. Works perfectly fine, but adding a second config file isn't ideal

**Comment by nzakas** [comment]
**TSC Summary:** This issue proposes adding a way for fine-grained disabling of inline directives on a per-rule basis. We currently have `noInlineConfig` that disables all inline configuration, but this request is for configuring based on rule name, so you could allow inline configuration of some rules but not others.

**TSC Question:** Do we want to support this idea and go to the next step of creating an RFC?

**Comment by Brian-Bartels** [comment]
Just let me know how I can help. Happy to make a RFC and or a draft PR. I would probably need some pointers to documentation about the new eslintrc architecture and how I could integrate into that

**Comment by btmills** [comment]
After discussing this in the TSC meeting, we're open to an RFC to explore how much complexity this might add. @Brian-Bartels you can find RFC instructions in the [RFC repository's readme](https://github.com/eslint/rfcs/), and that repository also contains the design document for the [new config system](https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification).

**Comment by Brian-Bartels** [comment]
Just pushed the RFC. Not sure if it's up to spec. Let me know if I'm missing details. Thanks :)

**Comment by Brian-Bartels** [comment]
@joshwilsonvu I see that you thumbs downed this suggestion. I encourage you to go to the RFC PR mentioned above and leave comments. I haven't sunk too much time into this and would be super interested in making sure that whatever I implement aligns with the community. The design has changed a fair amount from the above post. But the spirit is still the same

**Comment by joshwilsonvu** [comment]
Hi, sorry if that came across harshly, let me explain what I was thinking. My main thought is that I don't think ESLint should go farther than rules like `eslint-comments/no-restricted-disable`. If the purpose of lint rules is to define and maintain a set of standards that an organization adheres to, then the purpose of `eslint-disable` directives is to signal exceptions to these standards.

These exceptions should be red flags in code reviews, and every disable directive should have clear & obvious reasoning behind it. If it becomes normal to disable a specific rule, then the organization should probably turn the rule off, or find and fix all of these occurences if the rule is deemed important. I'm sure you agree, more or less. As it relates to your scenario with mandated rules, as I understand it the community solution forces you to write 

```js
// eslint-disable-next-line eslint-comments/no-restricted-disable
// eslint-disable-next-lline foo/some-restricted-rule
forbidden(code);
```

to disable `foo/some-restricted-rule`. I honestly sympathize with your frustration that code like this is written and gets checked in, but if a red flag like this slips past code review, what's to stop this code

```diff
"rules": {
- "foo/some-restricted-rule": "forbid"
+ "foo/some-restricted-rule": "error"
}
```

from getting in unnoticed? It feels like a technical band-aid to an organizational problem.

The benefit of the community solution is that you can grep for it, as you've seen for yourself: `eslint-comments/no-restricted-disable` or your rule of choice will be there every time this happens. It should be possible to get the number of these comments down to zero with some work, without breaking your build in the meantime. Afterwards, if you're still concerned about missing these double-disable-directives, it may just be simplest to grep for these from time to time and instruct PR reviewers not to pass PRs with them. You'd have to check for changes in `.eslintrc` from time to time anyway.

With all due respect for the tough situation you're in, I'm just a little doubtful how much value there is in adding `"forbid"`, when in my opinion ESLint is already doing its job—automatically maintaining standards, and leaving exceptions for human review.

**Comment by Brian-Bartels** [comment]
hi @joshwilsonvu no hard feelings whatsoever. I was just hoping to solicit feedback as this is the first large change that I've looked into making in ESLint.

To address your concerns though. I agree 100%. This is a band-aid over what I would describe as human fallibility. However, in my professional career, I've only ever worked on repos that have had at least 1000 developers coworking. And unfortunately, either because people aren't doing a proper review, or just because the PRs are so huge as to miss things, these disable directives slip through the cracks. It would be great to programmatically prevent them as described above. It is however, fairly easy to prevent people from adjusting the config by adding required reviewers automatically to those files or folders that would be able to adjust those configurations. As such I believe that adding the ability to specify rules that cannot be changed via inline configurations will be a strong barrier against rotten egg PRs getting through.

For what it's worth, the design has changed a fair amount from the original issue post above. It's much less obtrusive and will be easy to set up. But also easy to ignore for the many repos where this solution does Not make sense.

**Comment by joshwilsonvu** [comment]
I will say that most of my thoughts were along the lines of "this shouldn't be necessary" but not "this isn't practical," especially for such a large number of contributors. I've never experienced anything close to that 😅 You raise a good point about adding required reviewers, I didn't know about that.

I do see how this could be quite useful in a case like yours; where manual processes haven't yielded the results you're after, an automated tool can provide the strictness you need. If the ESLint team does decide to move forward with it, I hope that the language around the feature is clear that `"forbid"` shouldn't be reached for very often, and the behavior with shared configs is considered carefully. (I don't want this popping up in common starter configs!) For what it's worth I do think it's a nice design for its purpose!

**Comment by Brian-Bartels** [comment]
The updated design drops the term `forbid` entirely. In favor of making `noInlineConfig` accept an array as well as a boolean. This means that developers would have to know what they're getting into before enabling this updated feature. It also means that it won't be evangelized as a common use case. But more of an edge case (which it is)

**Comment by mdjermanovic** [comment]
> @mdjermanovic One concern with that workaround. How would it integrate into VSCode? I worry that people would disable a rule expecting it to pass PR build step. But then we'd sneak in a second run of ESlint and fail their builds unexpectedly.
> 
> I do also think that that workaround falls into the "slightly hacky" category. Works perfectly fine, but adding a second config file isn't ideal

I agree that a workaround with two config files and running ESLint twice isn't ideal, but to answer your question about VSCode, for the case that someone will use this workaround, and also because VSCode error reporting is mentioned [in the RFC](https://github.com/Brian-Bartels/rfcs/blob/494668f26904b57084753019224489f77d8dbee7/designs/2022-forbid-disable/README.md?plain=1#L75) as a drawback for this alternative.

In the main config file, in `extends: []` you can add a relative path to the other config file.

```js
//--------- .eslintrc.js -----------

module.exports = {
    extends: ["./.eslintrc2.js"],

    // .... main configuration
};
```

```js
//--------- .eslintrc2.js -----------

module.exports = {
    plugins: ["eslint-comments"],
    rules: {
        "eslint-comments/no-restricted-disable": ["error", "no-undef" /*, ... */]
    }
};
```



This way, VSCode and the main run will have the `eslint-comments/no-restricted-disable` rule enabled, and they'll report "Disabling 'no-undef' is not allowed" when the `no-undef` is disabled with a comment directive. If the contributor nevertheless decides to disable `eslint-comments/no-restricted-disable` as well, the second run of ESLint on CI with `npx eslint . --no-inline-config --no-eslintrc -c .eslintrc2.js` will catch that.

**Comment by Brian-Bartels** [comment]
@mdjermanovic That's a good callout. I do want to point out though that it only approximates the actual CI gate. Because VScode will suggest that they disable the rule. And when they do (¯\\\_(ツ)\_/¯) It'll clear the error

**Comment by nzakas** [comment]
There is an RFC open for this issue.

**Comment by nzakas** [comment]
Not to repeat myself, but there’s an RFC open for this issue.

**Comment by btmills** [comment]
To quote nzakas, there's an RFC open for this issue.

**Comment by ljharb** [comment]
not to repeat comments, but $repetition

**Comment by nzakas** [comment]
I believe by assigning someone to the issue it will appease the bot.

**Comment by nwcm** [comment]
likewise the ability to disable top level disable comments is also valuable. Restricts the scope of potential bad allows

**Comment by nzakas** [comment]
As an update, the RFC for this issue was closed because feedback was never incorporated. If anyone would like to revive the RFC, here is the original:
https://github.com/eslint/rfcs/pull/86

**Comment by AndreiRailean** [comment]
This is a good rule to have. When combined with CODEOWNERS, it would make it impossible to bypass certain rules without triggering additional reviews. So if `eslintrc` files are owned by the codeops team, then you wouldn't be able to bypass a rule without getting them to approve the global rule change.

**Comment by nzakas** [comment]
@AndreiRailean thanks, we aren't looking for more opinions on this. If you like the idea, please consider creating an RFC for it. That's how to move this issue forward.

**Comment by mderazon** [comment]
Only tangibly related but as a middle ground, can anyone know of a tool to have a PR with a new `eslint-disable-next-line` to require CODEOWNERS approval or to fail the build ?

**Comment by ljharb** [comment]
@mderazon https://www.npmjs.com/package/eslint-plugin-eslint-comments

**Comment by nzakas** [comment]
Closing due to age.
